assets/js/blocks/stock-filter/block.js

```diff
@@ -0,0 +1,285 @@
+	] = useQueryStateByKey( 'stock_status', [] );
+
+	const {
+		results: filteredCounts,
+		isLoading: filteredCountsLoading,
+	} = useCollectionData( {
+		queryStock: true,
+		queryState,
+	} );
+
+	/**
+	 * Get count data about a given status by slug.
+	 */
+	const getFilteredStock = useCallback(
+		( slug ) => {
+			if ( ! filteredCounts.stock_status_counts ) {
+				return null;
+			}
+			return filteredCounts.stock_status_counts.find(
+				( { status, count } ) =>
+					status === slug && Number( count ) !== 0
+			);
+		},
+		[ filteredCounts ]
+	);
+
+	/**
+	 * Compare intersection of all stock statuses and filtered counts to get a list of options to display.
+	 */
+	useEffect( () => {
+		/**
+		 * Checks if a status slug is in the query state.
+		 *
+		 * @param {string} queryStatus The status slug to check.
+		 */
+		const isStockStatusInQueryState = ( queryStatus ) => {
+			if ( ! queryState?.stock_status ) {
+				return false;
+			}
+			return queryState.stock_status.some( ( { status = [] } ) =>
+				status.includes( queryStatus )
+			);
+		};
+
+		if ( filteredCountsLoading || blockAttributes.isPreview ) {
+			return;
+		}
+
+		const newOptions = initialOptions
+			.map( ( status ) => {
+				const filteredStock = getFilteredStock( status.slug );
+
+				if (
+					! filteredStock &&
+					! checked.includes( status.slug ) &&
+					! isStockStatusInQueryState( status.slug )
+				) {
+					return null;
+				}
+
+				const count = filteredStock ? Number( filteredStock.count ) : 0;
+
+				return {
+					value: status.slug,
+					name: decodeEntities( status.name ),
+					label: (
+						<Label
+							name={ decodeEntities( status.name ) }
+							count={ blockAttributes.showCounts ? count : null }
+						/>
+					),
+				};
+			} )
+			.filter( Boolean );
+
+		setDisplayedOptions( newOptions );
+	}, [
+		blockAttributes.showCounts,
+		blockAttributes.isPreview,
+		filteredCountsLoading,
+		getFilteredStock,
+		checked,
+		queryState.stock_status,
+	] );
+
+	const onSubmit = useCallback(
+		( isChecked ) => {
+			if ( isEditor ) {
+				return;
+			}
+			if ( isChecked ) {
+				setProductStockStatusQuery( checked );
+			}
+		},
+		[ isEditor, setProductStockStatusQuery, checked ]
+	);
+
+	// Track checked STATE changes - if state changes, update the query.
+	useEffect( () => {
+		if ( ! blockAttributes.showFilterButton ) {
+			onSubmit( checked );
+		}
+	}, [ blockAttributes.showFilterButton, checked, onSubmit ] );
+
+	const checkedQuery = useMemo( () => {
+		return productStockStatusQuery;
+	}, [ productStockStatusQuery ] );
+
+	const currentCheckedQuery = useShallowEqual( checkedQuery );
+	const previousCheckedQuery = usePrevious( currentCheckedQuery );
+	// Track Stock query changes so the block reflects current filters.
+	useEffect( () => {
+		if (
+			! isShallowEqual( previousCheckedQuery, currentCheckedQuery ) && // checked query changed
+			! isShallowEqual( checked, currentCheckedQuery ) // checked query doesn't match the UI
```

Comment:
Making a suggestion to update these comments to use sentence case 🙂 

```suggestion
			! isShallowEqual( previousCheckedQuery, currentCheckedQuery ) && // Checked query changed.
			! isShallowEqual( checked, currentCheckedQuery ) // Checked query doesn't match the UI.
```

Reply:
Fixed!